pkg/registry/core/pod/rest/subresources.go

```diff
@@ -156,7 +172,20 @@ func (r *ExecREST) Connect(ctx context.Context, name string, opts runtime.Object
 	if err != nil {
 		return nil, err
 	}
-	return newThrottledUpgradeAwareProxyHandler(location, transport, false, true, responder), nil
+	handler := newThrottledUpgradeAwareProxyHandler(location, transport, false, true, responder)
+	if utilfeature.DefaultFeatureGate.Enabled(features.ClientRemoteCommandWebsockets) {
+		// Wrap the upgrade aware handler to implement stream translation
+		// for WebSocket/V5 upgrade requests.
+		streamOptions := translator.Options{
+			Stdin:  execOpts.Stdin,
+			Stdout: execOpts.Stdout,
+			Stderr: execOpts.Stderr,
+			Tty:    execOpts.TTY,
+		}
+		streamtranslator := translator.NewStreamTranslatorHandler(location, transport, proxy.NewErrorResponder(responder), streamOptions)
```

Comment:
I'm really surprised the stream translator needs `location` and `transport`... I expected it to be translating at the `http.Request/http.ResponseWriter` level and delegating the kubelet transport stuff to the existing `throttledUpgradeAwareProxyHandler` which already knows how to talk to the kubelet

Reply:
the signature I expected here was 

```go
streamtranslator := translator.NewStreamTranslatorHandler(handler, streamOptions)`
```

Reply:
this object will be temporary and eventually removed when we have all path implemented, right?

Reply:
I don't think we'll ever be able to drop this, since it'll be really hard to know when all CRIs support the v5 protocol

I dug more and the reason we can't pass in the handler and wire the spdy client up to use it is that the spdy client needs the dialer/connection, not just the round tripper (req/resp).

Reply:
Several things jump out at me that the existing handler is doing we need to ensure happen with this translator if it's going to directly connect to kubelets:

1. throttling behavior
2. not following redirects
    * https://github.com/kubernetes/kubernetes/pull/95128
    * https://github.com/kubernetes/kubernetes/pull/112193
    * https://github.com/kubernetes/kubernetes/pull/112526
3. timeout handling
    * https://github.com/kubernetes/kubernetes/pull/71419
4. verifying non-upgrade responses terminate the connection
    * https://github.com/kubernetes/kubernetes/pull/71412
    * https://github.com/kubernetes/kubernetes/pull/92941

Reply:
most of them are not exercised if the connection is upgraded, no? it seems the handler returns on `tryUprade`

https://github.com/kubernetes/kubernetes/blob/6d78a25374d064546bd902b99d4dc1d3f7d5c7c2/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L212-L216

This seems the ones missing

https://github.com/kubernetes/kubernetes/pull/92941
https://github.com/kubernetes/kubernetes/pull/71419

Reply:
I think @aojea is correct. The `redirect` items you have enumerated are not relevant, since all the redirect fixes added to the `UpgradeAwareHandler` occur in the non-upgrade portion (ie. they do not happen in `tryUpgrade`).

More Context: the RemoteCommand suprotocol was updated in version four (`v4.channel.k8s.io`) to add an error stream. The initial task in the `StreamTranslator.ServeHTTP()` is to upgrade the HTTP request to a WebSocket connection (including the error stream). So if this initial upgrade succeeds, the error stream is available to return a structured error.

Reply:
> I don't think we'll ever be able to drop this, since it'll be really hard to know when all CRIs support the v5 protocol
> 
> I dug more and the reason we can't pass in the handler and wire the spdy client up to use it is that the spdy client needs the dialer/connection, not just the round tripper (req/resp).

The plan is to extend the `StreamTranslator` to the kubelet, therefore implementing all intra-cluster streaming communication with WebSockets (e.g. only the node communication between the kubelet and the container runtime through the CRI would use SPDY).